File: CN Lab/18501A0591_4.c
/*distance vector routing */
#include<stdio.h>
struct node
{
int dis[20];
int from[20];
}rt[10];
int main()
{
    int n;
    printf("Enter Number Of Nodes : ");
    scanf("%d",&n);
    int costmat[n][n],i,j,k;
    printf("Enter Adjacency Matrix : \n");
    for(i=0;i<n;i++)
    {
        for(j=0;j<n;j++)
        {
            scanf("%d",&costmat[i][j]);
            costmat[i][i]=0;
            rt[i].from[j]=j;
            rt[i].dis[j]=costmat[i][j];
        }
    }
    int itr=1,count=0;
    do
    {
        count=0;
        printf("iteration : %d",itr++);
        for(i=0;i<n;i++)
        {
            printf("\n\n For router %c \n",i+97);
            printf("\nDestination\tNext Hop\tDistance\n");
            for(j=0;j<n;j++)
            {
                printf("%c\t\t%c\t\t%d\n",i+97,j+97,rt[i].dis[j]);
            }
        }
        for(i=0;i<n;i++)
        {
            for(j=0;j<n;j++)
            {
                for(k=0;k<n;k++)
                {
                    if(rt[i].dis[j]>costmat[i][k]+rt[k].dis[j])
                    {
                        rt[i].dis[j]=costmat[i][k]+rt[k].dis[j];
                        rt[i].from[j]=k;
                        count=1;
                    }
                }
            }
        }
    }while(count!=0);
    return 0;
}

File: CN Lab/18501A0591_4.py
#18501A0591_4.py
class routing_table:
    max_num = (1<<31)-1
    d = {'dist':max_num,'next_hop':-1}
    def __init__(self, n):
        #initiating dist to max and next hop -1
        self.n = n
        self.dest = [{'dist':(1<<31)-1,'next_hop':-1} for i in range(n)] 

    def print1(self):
        for i in range(self.n):
            print(i,self.dest[i]['dist'],self.dest[i]['next_hop'])
        print()

class graph:
    def __init__(self,n):
        self.n = n
        self.nodes = [routing_table(n) for i in range(n)]
        print('Adjacency matrix:')
        self.adj_mat = [list(map(int,input().split())) for i in range(n)]

    def first_change(self):
        nodecount = 0
        for node in self.nodes:
            for i in range(self.n):
                node.dest[i]['dist'] = self.adj_mat[nodecount][i]
                node.dest[i]['next_hop'] = i
                #self.print1()
            nodecount+=1
        return True
    

    def find_next_hops(self):
        flag = False
        for node in self.nodes:
            #for every dest "i" in "node"'s routing table find next min possible hop
            for i in range(self.n):
                for j in range(self.n):
                    via_j = node.dest[j]['dist'] + (self.adj_mat[j][i] if self.adj_mat[i][j]!=0 else (1<<31)-1)
                    if node.dest[i]['dist']>via_j:
                        node.dest[i]['dist'] = via_j
                        node.dest[i]['next_hop'] = j
                        flag = True
        return flag


    def distance_vector_routing(self):
        changed = self.first_change()
        hop_count = 1
        print('hop{}:'.format(hop_count))
        self.print1()
        while changed:
            hop_count+=1
            changed = self.find_next_hops()
            print('hop{}'.format(hop_count))

    def print1(self):
        for node in range(self.n):
            print('node{}-dist-nextHop'.format(node))
            self.nodes[node].print1()
            





n = int(input('Enter no.of nodes:'))

#creating graph
g = graph(n)

g.distance_vector_routing()
print()
g.print1()


